### ISSUE TITLE
Api Proposal: Stats/Callback when PipeWriter threshold is reached

### ISSUE THREAD

**Issue by rohits79** [dotnet/corefx#42406]
For slow consumers, it would be good to detect that there is potential back pressure building up i.e. pipe reader is not advancing fast enough. Would be good see some statistics, or callback so we could debug and analyze.

**Comment by halter73** [comment]
We've already deprecated PipeWriter.OnReaderCompleted (#38362) in part because it makes it makes implementing non-default PipeWriters harder. Same goes for PipeReader.OnWriterCompleted.

I think this proposal faces similar problems as those callbacks. Fortunately, since PipeWriter is abstract, you can do something similar to what was suggested in https://github.com/dotnet/corefx/issues/38362#issuecomment-511870173 and decorate your PipeWriter with a custom version that does fire a callback when PipeWriter.FlushAsync/WriteAsync returns and incomplete ValueTask.

**Comment by FiniteReality** [comment]
For metrics, having a BytesWritten/BytesRead would be interesting - just like #39998 and #30852. When debugging code, it's likely easier to see "Oh, this pipeline has a bunch of data backed up and nobody's processing it." - I had a few issues like this myself and looking at the internal properties `_unconsumedBytes` and `_unflushedBytes` helped identify a few places where I was missing FlushAsync calls or my ReadAsync code had thrown an exception which was swallowed elsewhere

**Comment by halter73** [comment]
> For metrics, having a BytesWritten/BytesRead would be interesting

Agreed. There's already been some discussion about this at https://github.com/dotnet/corefx/issues/31074 and elsewhere.
